### PR TITLE
add oneflow backend for mmeval

### PR DESCRIPTION
**Pull request by ofhwei**
## Motivation

add oneflow backend for mmeval.

## Modification
add the following metrics and unittests in oneflow:
- accuracy 
- end_point_error 
- f_metric 
- mean_iou 
- multi_label 
- single_label

### CODE REVIEW

File: mmeval/metrics/accuracy.py [hunk 1]
```diff
@@ -13,10 +13,12 @@
     import tensorflow
     import tensorflow as tf
     import torch
+    import oneflow as flow
 else:
     paddle = try_import('paddle')
     torch = try_import('torch')
     tf = try_import('tensorflow')
+    flow = try_import("oneflow")
```
Comment (zhouzaida) [inline mmeval/metrics/accuracy.py:21]:
```suggestion
    flow = try_import(oneflow')
```

File: mmeval/metrics/accuracy.py [hunk 2]
```diff
@@ -29,8 +31,8 @@ def _is_scalar(obj: np.number):  # type: ignore
 @overload
 @dispatch
 def _is_scalar(obj: Union[np.ndarray,  # type: ignore
-                          'torch.Tensor', 'tensorflow.Tensor']):
-    """Check if a ``np.ndarray`` | ``torch.Tensor`` | ``tensorflow.Tensor`` is
+                          'torch.Tensor', 'oneflow.Tensor', 'tensorflow.Tensor']):
```
Comment (zhouzaida) [inline mmeval/metrics/accuracy.py:34]:
```suggestion
                          'torch.Tensor', 'flow.Tensor', 'tensorflow.Tensor']):
```

File: mmeval/metrics/accuracy.py [hunk 3]
```diff
@@ -51,6 +53,11 @@ def _torch_topk(inputs: 'torch.Tensor',
     """Invoke the PyTorch topk."""
     return inputs.topk(k, dim=dim)
 
+def _oneflow_topk(inputs: 'oneflow.Tensor',
```
Comment (zhouzaida) [inline mmeval/metrics/accuracy.py:56]:
```suggestion
def _oneflow_topk(inputs: 'flow.Tensor',
```

File: mmeval/metrics/accuracy.py [hunk 4]
```diff
@@ -87,7 +98,7 @@ class Accuracy(BaseMetric):
     This metric computes the accuracy based on the given topk and thresholds.
 
     Currently, this metric supports 4 kinds of inputs, i.e. ``numpy.ndarray``,
-    ``torch.Tensor``, ``tensorflow.Tensor`` and ``paddle.Tensor``, and the
+    ``torch.Tensor``, ``OneFlow.Tensor``, ``tensorflow.Tensor`` and ``paddle.Tensor``, and the
```
Comment (zhouzaida) [inline mmeval/metrics/accuracy.py:101]:
```suggestion
    ``torch.Tensor``, ``oneflow.Tensor``, ``tensorflow.Tensor`` and ``paddle.Tensor``, and the
```

File: mmeval/metrics/accuracy.py [hunk 5]
```diff
@@ -225,6 +236,58 @@ def _compute_corrects(
                     0, keepdim=True).float()
         return corrects_per_sample
 
+    @overload  # type: ignore
+    @dispatch
+    def _compute_corrects(
+        self, predictions: Union['oneflow.Tensor', Sequence['oneflow.Tensor']],
+        labels: Union['oneflow.Tensor',
+                      Sequence['oneflow.Tensor']]) -> 'oneflow.Tensor':
```
Comment (zhouzaida) [inline mmeval/metrics/accuracy.py:244]:
```suggestion
    def _compute_corrects(
        self, predictions: Union['flow.Tensor', Sequence['flow.Tensor']],
        labels: Union['flow.Tensor',
                      Sequence['flow.Tensor']]) -> 'flow.Tensor':
```

File: tests/test_metrics/test_multi_label.py [hunk 6]
```diff
@@ -149,6 +168,30 @@ def test_metric_interface_torch(metric_kwargs, preds, labels):
...
+    # one-hot encodings in Sequence
+    [flow.Tensor([1, 0, 0]), flow.Tensor([0, 1, 0])],
+    [flow.Tensor([0]), flow.Tensor([1])],  # label-format in Sequence
+])
+@pytest.mark.skipif(flow is None, reason='OneFlow is not available!')
+def test_metric_interface_oneflow(metric_kwargs, preds, labels):
+    """Test all kinds of inputs in OneFlow."""
+    multi_label_metric = MultiLabelMetric(**metric_kwargs)
+    results = multi_label_metric(preds, labels)
+    assert isinstance(results, dict)
```
Comment (ice-tong) [inline tests/test_metrics/test_multi_label.py:192]:
```suggestion
@pytest.mark.parametrize('metric_kwargs', [{'num_classes': 3}])
@pytest.mark.parametrize('preds', [
    [[0.1, 0.9, 0.8], [0.5, 0.6, 0.8]],  # scores in a Tensor
    # scores in Sequence
    [[0.1, 0.9, 0.8], [0.5, 0.6, 0.8]],
    [[0, 1, 0], [1, 1, 0]],  # one-hot encodings in a Tensor
    # one-hot encodings in Sequence
    [[0, 1, 0], [1, 1, 0]],
    # [[1], [0, 1]],  # label-format in Sequence
])
@pytest.mark.parametrize('labels', [
    [[1, 0, 0], [0, 1, 0]],  # one-hot encodings in a Tensor
    # one-hot encodings in Sequence
    [[1, 0, 0], [0, 1, 0]],
    [[0], [1]],  # label-format in Sequence
])
@pytest.mark.skipif(flow is None, reason='OneFlow is not available!')
def test_metric_interface_oneflow(metric_kwargs, preds, labels):
    """Test all kinds of inputs in OneFlow."""
    multi_label_metric = MultiLabelMetric(**metric_kwargs)
    results = multi_label_metric(flow.Tensor(preds), flow.Tensor(labels))
    assert isinstance(results, dict)
```

File: tests/test_metrics/test_multi_label.py [hunk 7]
```diff
@@ -102,6 +101,26 @@ def test_metric_interface_torch_topk(metric_kwargs, preds, labels):
...
+    # one-hot encodings in Sequence
+    [flow.Tensor([1, 0, 0]), flow.Tensor([0, 1, 0])],
+    [flow.Tensor([0]), flow.Tensor([1])],  # label-format in Sequence
+])
+@pytest.mark.skipif(flow is None, reason='OneFlow is not available!')
+def test_metric_interface_oneflow_topk(metric_kwargs, preds, labels):
+    """Test scores inputs with topk in OneFlow."""
+    multi_label_metric = MultiLabelMetric(**metric_kwargs)
+    results = multi_label_metric(preds, labels)
+    assert isinstance(results, dict)
```
Comment (ice-tong) [inline tests/test_metrics/test_multi_label.py:121]:
```suggestion
@pytest.mark.parametrize('metric_kwargs', [{'num_classes': 3, 'topk': 1}])
@pytest.mark.parametrize('preds', [
    [[0.1, 0.9, 0.8], [0.5, 0.6, 0.8]],  # scores in a Tensor
    # scores in Sequence
    [[0.1, 0.9, 0.8], [0.5, 0.6, 0.8]],
])
@pytest.mark.parametrize('labels', [
    [[1, 0, 0], [0, 1, 0]],  # one-hot encodings in a Tensor
    # one-hot encodings in Sequence
    [[1, 0, 0], [0, 1, 0]],
    [[0], [1]],  # label-format in Sequence
])
@pytest.mark.skipif(flow is None, reason='OneFlow is not available!')
def test_metric_interface_oneflow_topk(metric_kwargs, preds, labels):
    """Test scores inputs with topk in OneFlow."""
    multi_label_metric = MultiLabelMetric(**metric_kwargs)
    results = multi_label_metric(flow.Tensor(preds), flow.Tensor(labels))
    assert isinstance(results, dict)
```

File: mmeval/metrics/multi_label.py [hunk 8]
```diff
@@ -389,6 +405,38 @@ def _compute_metric(self, predictions: Sequence['torch.Tensor'],
         return _precision_recall_f1_support(  # type: ignore
             pos_inds, labels, self.average)
 
+    @overload  # type: ignore
+    @dispatch
+    def _compute_metric(  # type: ignore
+            self, predictions: Sequence['oneflow.Tensor'],
+            labels: Sequence['oneflow.Tensor']) -> List:
+        """A OneFlow implementation that computes the metric."""
```
[the remark this answers is not in the document]
Comment (ofhwei) [reply mmeval/metrics/multi_label.py:413]: I'm not sure , but look [here]( https://strategiesforparents.com/a-one-or-an-one-understanding-correct-grammar/ ) :)

File: tests/test_metrics/test_multi_label.py (hunk 6 above)
Comment (ofhwei) [reply tests/test_metrics/test_multi_label.py:192]: This is borrowed from [here](https://github.com/open-mmlab/mmeval/blob/main/tests/test_metrics/test_multi_label.py#L145), i think it would be better to keep the same style.

Comment (zhouzaida) [reply tests/test_metrics/test_multi_label.py:192]: Those unit tests will be refactored because they will fail if torch is not installed.

File: mmeval/metrics/end_point_error.py [hunk 9]
```diff
@@ -146,6 +149,31 @@ def end_point_error_map(
...
+        label: 'oneflow.Tensor',
+        valid_mask: Optional['oneflow.Tensor'] = None) -> Tuple[np.ndarray,
+                                                                int]:
+        """Calculate end point error map.
+
+        Args:
+            prediction (oneflow.Tensor): Prediction with shape (H, W, 2).
+            label (oneflow.Tensor): Ground truth with shape (H, W, 2).
+            valid_mask (oneflow.Tensor, optional):
+                Valid mask with shape (H, W).
```
Comment (zhouzaida) [inline mmeval/metrics/end_point_error.py:165]:
```suggestion
            valid_mask (oneflow.Tensor, optional): Valid mask with
                shape (H, W).
```

File: mmeval/metrics/multi_label.py [hunk 10]
```diff
@@ -10,32 +10,42 @@
...
-        label (torch.Tensor or np.ndarray): The label-format input.
-            The format of item must be label-format.
+        label (torch.Tensor or oneflow.Tensor or np.ndarray):
+            The label-format input. The format of item must be label-format.
         num_classes (int): The number of classes.
 
     Return:
-        torch.Tensor or np.ndarray: The converted one-hot encodings.
+        torch.Tensor or oneflow.Tensor or np.ndarray:
+            The converted one-hot encodings.
```
Comment (zhouzaida) [inline mmeval/metrics/multi_label.py:41]:
```suggestion
        torch.Tensor or oneflow.Tensor or np.ndarray:
        The converted one-hot encodings.
```

File: mmeval/metrics/multi_label.py [hunk 11]
```diff
@@ -46,25 +56,31 @@ def label_to_onehot(label: Union[np.ndarray, 'torch.Tensor'],
...
+        data (Union[Sequence[np.ndarray, 'torch.Tensor', 'oneflow.Tensor'],
+        np.ndarray, 'torch.Tensor', 'oneflow.Tensor']):
+            The input data of prediction or labels.
         num_classes (int): The number of classes.
         is_onehot (bool): Whether the data is one-hot encodings.
 
     Return:
-        torch.Tensor or np.ndarray: One-hot encodings or predict scores.
+        torch.Tensor or oneflow.Tensor or np.ndarray:
+            One-hot encodings or predict scores.
```
Comment (zhouzaida) [inline mmeval/metrics/multi_label.py:78]:
```suggestion
        torch.Tensor or oneflow.Tensor or np.ndarray:
        One-hot encodings or predict scores.
```

File: mmeval/metrics/multi_label.py [hunk 12]
```diff
@@ -721,19 +836,21 @@ def _compute_metric(
...
         This method would be invoked in `BaseMetric.compute` after distributed
         synchronization.
         Args:
-            results (List[Union[NUMPY_IMPL_HINTS, TORCH_IMPL_HINTS]]): A list
-                of tuples that consisting the prediction and label. This list
-                has already been synced across all ranks.
+            results (List[Union[NUMPY_IMPL_HINTS, TORCH_IMPL_HINTS,
+            ONEFLOW_IMPL_HINTS]]): A list of tuples that consisting the
+            prediction and label. This list has already been synced across
+            all ranks.
```
Comment (zhouzaida) [inline mmeval/metrics/multi_label.py:853]:
```suggestion
            all ranks.

```

**Review by zhouzaida — APPROVED** [review]

**Comment by zhouzaida** [comment]
#59 has been merged and please update this PR with the main branch to trigger the CI tests.

File: mmeval/metrics/end_point_error.py [hunk 13]
```diff
@@ -122,13 +125,13 @@ def end_point_error_map(
...
-            label: 'torch.Tensor',
-            valid_mask: Optional['torch.Tensor'] = None
-    ) -> Tuple[np.ndarray, int]:
+    @dispatch  # type: ignore
+    def end_point_error_map(  # type: ignore
+        self,
+        prediction: 'torch.Tensor',
+        label: 'torch.Tensor',
+        valid_mask: Optional['torch.Tensor'] = None
+    ) -> Tuple[np.ndarray, int]:  # type: ignore
```
Comment (ice-tong) [inline mmeval/metrics/end_point_error.py:134]: We should add `typing.overload` decorator to this function and remove unnecessary changes.
```suggestion
    @overload
    @dispatch  # type: ignore
    def end_point_error_map(
            self,
            prediction: 'torch.Tensor',
            label: 'torch.Tensor',
            valid_mask: Optional['torch.Tensor'] = None
    ) -> Tuple[np.ndarray, int]:
```

File: mmeval/metrics/multi_label.py [hunk 14]
```diff
@@ -721,19 +836,22 @@ def _compute_metric(
...
         """Compute the metric.
 
-        Currently, there are 2 implementations of this method: NumPy and
-        PyTorch. Which implementation to use is determined by the type of the
-        calling parameters. e.g. `numpy.ndarray` or `torch.Tensor`.
+        Currently, there are 3 implementations of this method: NumPy and
+        PyTorch and OneFlow. Which implementation to use is determined by the
+        type of the calling parameters. e.g. `numpy.ndarray` or
+        `torch.Tensor`, `oneflow.Tensor`.
         This method would be invoked in `BaseMetric.compute` after distributed
```
Comment (ice-tong) [inline mmeval/metrics/multi_label.py:847]: Need a blank line before this line

File: mmeval/metrics/multi_label.py [hunk 15]
```diff
@@ -721,19 +836,22 @@ def _compute_metric(
...
         synchronization.
         Args:
```
Comment (ice-tong) [inline mmeval/metrics/multi_label.py:849]: Need a blank line before this line

File: tests/test_metrics/test_single_label_metric.py [hunk 16]
```diff
@@ -51,6 +50,21 @@ def test_metric_torch_assertion():
             torch.Tensor([[0.1, 0.9, 0], [0.5, 0.5, 0]]), torch.Tensor([0, 1]))
 
 
+@pytest.mark.skipif(flow is None or flow.__version__ < '0.8.1',
+                    reason='OneFlow > 0.8.0 is required!')
```
Comment (ice-tong) [inline tests/test_metrics/test_single_label_metric.py:54]: String use LooseVersion for version comparison.
```
from distutils.version import LooseVersion
```

**Review by ice-tong — APPROVED** [review]
LGTM with 4 comments
